include/picongpu/fields/background/templates/twtstight/TWTSTight.tpp

```diff
@@ -0,0 +1,366 @@
+                tdelay = detail::getInitialTimeDelay_SI(
+                    auto_tdelay,
+                    tdelay_user_SI,
+                    halfSimSize,
+                    pulselength_SI,
+                    focus_y_SI,
+                    phi,
+                    beta_0);
+                if(phi < 0.0_X)
+                    phiPositive = float_X(-1.0);
+            }
+
+            template<typename T_Field>
+            HDINLINE float3_X TWTSTight<T_Field>::getTWTSField_Normalized(
+                pmacc::math::Vector<floatD_64, detail::numComponents> const& fieldPositions_SI,
+                float_64 const time) const
+            {
+                using PosVecVec = pmacc::math::Vector<float3_64, detail::numComponents>;
+                PosVecVec pos(PosVecVec::create(float3_64::create(0.0)));
+
+                if constexpr(simDim == DIM3)
+                {
+                    for(uint32_t k = 0; k < detail::numComponents; ++k)
+                    {
+                        for(uint32_t i = 0; i < simDim; ++i)
+                        {
+                            pos[k][i] = fieldPositions_SI[k][i];
+                        }
+                    }
+                    /* E- or B-field normalized to the peak amplitude. */
+                    return static_cast<float3_X>(
+                        calcTWTSFieldX(pos[0], time),
+                        calcTWTSFieldY(pos[1], time),
+                        calcTWTSFieldZ(pos[2], time));
+                }
+                if constexpr(simDim == DIM2)
+                {
+                    for(uint32_t k = 0; k < detail::numComponents; ++k)
+                    {
+                        /* 2D (y,z) vectors are mapped on 3D (x,y,z) vectors. */
+                        for(uint32_t i = 0; i < simDim; ++i)
+                        {
+                            pos[k][i + 1] = fieldPositions_SI[k][i];
+                        }
+                    }
+                    /* E- or B-field normalized to the peak amplitude. */
+                    return static_cast<float3_X>(
+                        calcTWTSFieldX(pos[0], time),
+                        calcTWTSFieldY(pos[1], time),
+                        calcTWTSFieldZ(pos[2], time));
+                }
+                // We should never be here.
+                else
+                    return static_cast<float3_X>(NAN);
+            }
+
+            template<typename T_Field>
+            HDINLINE float3_X
+            TWTSTight<T_Field>::operator()(DataSpace<simDim> const& cellIdx, uint32_t const currentStep) const
+            {
+                traits::FieldPosition<fields::YeeCell, T_Field> const fieldPos;
+                return getValue(precisionCast<float_X>(cellIdx), fieldPos(), static_cast<float_X>(currentStep));
+            }
+
+            template<typename T_Field>
+            HDINLINE float3_X TWTSTight<T_Field>::operator()(floatD_X const& cellIdx, float_X const currentStep) const
```

Comment:
It is the required interface for incidentFields, which was defined in a general way, such that it can be extended to a whole range of possible field solvers. Similar as fractional cells it demands access to sub-timesteps.

Reply:
Thanks for the explanation.